docs/man/memstat.1
.TH "MEMSTAT" "1" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memstat \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
memstat \- Display the operating status of a single or group of memcached servers
.SH SYNOPSIS
.sp
memstat [options]
.SH DESCRIPTION
.sp
\fBmemstat\fP  dumps the state of memcached(1) servers.
It displays all data to stdout.
.SH OPTIONS
.sp
You can specify servers via the option:
.INDENT 0.0
.TP
.B \-\-servers
.UNINDENT
.sp
or via the environment variable:
.INDENT 0.0
.TP
.B \(aqMEMCACHED_SERVERS, \-\-args\(aq
.UNINDENT
.sp
which can be used to specify the "argument" sent to the stats command (ie slab, size, items, etc..).
.sp
For a full list of operations run the tool with:
.INDENT 0.0
.TP
.B \-\-help
.UNINDENT
.INDENT 0.0
.TP
.B \-\-analyze
.UNINDENT
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

clients/memstat.cc
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary:
 *
 * Authors: 
 *          Brian Aker
 *          Toru Maesaka
 */
#include <config.h>

#include <cstdio>
#include <cstring>
#include <ctime>
#include <iostream>
#include <fcntl.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/types.h>

#include <libmemcached/memcached.h>

#include "client_options.h"
#include "utilities.h"

#define PROGRAM_NAME "memstat"
#define PROGRAM_DESCRIPTION "Output the state of a memcached cluster."

/* Prototypes */
static void options_parse(int argc, char *argv[]);
static void run_analyzer(memcached_st *memc, memcached_stat_st *memc_stat);
static void print_analysis_report(memcached_st *memc,
                                  memcached_analysis_st *report);

static bool opt_binary= false;
static bool opt_verbose= false;
static bool opt_server_version= false;
static bool opt_analyze= false;
static char *opt_servers= NULL;
static char *stat_args= NULL;
static char *analyze_mode= NULL;

static struct option long_options[]=
{
  {(OPTIONSTRING)"args", required_argument, NULL, OPT_STAT_ARGS},
  {(OPTIONSTRING)"version", no_argument, NULL, OPT_VERSION},
  {(OPTIONSTRING)"help", no_argument, NULL, OPT_HELP},
  {(OPTIONSTRING)"quiet", no_argument, NULL, OPT_QUIET},
  {(OPTIONSTRING)"verbose", no_argument, NULL, OPT_VERBOSE},
  {(OPTIONSTRING)"binary", no_argument, NULL, OPT_BINARY},
  {(OPTIONSTRING)"debug", no_argument, NULL, OPT_DEBUG},
  {(OPTIONSTRING)"server-version", no_argument, NULL, OPT_SERVER_VERSION},
  {(OPTIONSTRING)"servers", required_argument, NULL, OPT_SERVERS},
  {(OPTIONSTRING)"analyze", optional_argument, NULL, OPT_ANALYZE},
  {0, 0, 0, 0},
};


static memcached_return_t stat_printer(memcached_server_instance_st instance,
                                       const char *key, size_t key_length,
                                       const char *value, size_t value_length,
                                       void *context)
{
  static memcached_server_instance_st last= NULL;
  (void)context;

  if (last != instance)
  {
    printf("Server: %s (%u)\n", memcached_server_name(instance),
           (uint32_t)memcached_server_port(instance));
    last= instance;
  }

  printf("\t %.*s: %.*s\n", (int)key_length, key, (int)value_length, value);

  return MEMCACHED_SUCCESS;
}

static memcached_return_t server_print_callback(const memcached_st *,
                                                memcached_server_instance_st instance,
                                                void *)
{
  std::cerr << memcached_server_name(instance) << ":" << memcached_server_port(instance) <<
    " " << int(memcached_server_major_version(instance)) << 
    "." << int(memcached_server_minor_version(instance)) << 
    "." << int(memcached_server_micro_version(instance)) << std::endl;

  return MEMCACHED_SUCCESS;
}

int main(int argc, char *argv[])
{
  options_parse(argc, argv);
  initialize_sockets();

  if (opt_servers == false)
  {
    char *temp;
    if ((temp= getenv("MEMCACHED_SERVERS")))
    {
      opt_servers= strdup(temp);
    }
    else
    {
      std::cerr << "No Servers provided" << std::endl;
      return EXIT_FAILURE;
    }
  }

  memcached_st *memc= memcached_create(NULL);
  memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, opt_binary);

  memcached_server_st *servers= memcached_servers_parse(opt_servers);
  free(opt_servers);

  memcached_return_t rc= memcached_server_push(memc, servers);
  memcached_server_list_free(servers);

  if (rc != MEMCACHED_SUCCESS and rc != MEMCACHED_SOME_ERRORS)
  {
    printf("Failure to communicate with servers (%s)\n",
           memcached_strerror(memc, rc));
    exit(EXIT_FAILURE);
  }

  if (opt_server_version)
  {
    if (memcached_failed(memcached_version(memc)))
    {
      std::cerr << "Unable to obtain server version";
      exit(EXIT_FAILURE);
    }

    memcached_server_fn callbacks[1];
    callbacks[0]= server_print_callback;
    memcached_server_cursor(memc, callbacks, NULL,  1);
  }
  else if (opt_analyze)
  {
    memcached_stat_st *memc_stat= memcached_stat(memc, NULL, &rc);

    if (memc_stat == NULL)
    {
      exit(EXIT_FAILURE);
    }

    run_analyzer(memc, memc_stat);

    memcached_stat_free(memc, memc_stat);
  }
  else
  {
    rc= memcached_stat_execute(memc, stat_args, stat_printer, NULL);
  }

  memcached_free(memc);

  return rc == MEMCACHED_SUCCESS ? EXIT_SUCCESS: EXIT_FAILURE;
}

static void run_analyzer(memcached_st *memc, memcached_stat_st *memc_stat)
{
  memcached_return_t rc;

  if (analyze_mode == NULL)
  {
    memcached_analysis_st *report;
    report= memcached_analyze(memc, memc_stat, &rc);
    if (rc != MEMCACHED_SUCCESS || report == NULL)
    {
      printf("Failure to analyze servers (%s)\n",
             memcached_strerror(memc, rc));
      exit(1);
    }
    print_analysis_report(memc, report);
    free(report);
  }
  else if (strcmp(analyze_mode, "latency") == 0)
  {
    uint32_t flags, server_count= memcached_server_count(memc);
    uint32_t num_of_tests= 32;
    const char *test_key= "libmemcached_test_key";

    memcached_st **servers;
    servers= static_cast<memcached_st**>(malloc(sizeof(memcached_st*) * server_count));
    if (not servers)
    {
      fprintf(stderr, "Failed to allocate memory\n");
      return;
    }

    for (uint32_t x= 0; x < server_count; x++)
    {
      memcached_server_instance_st instance=
        memcached_server_instance_by_position(memc, x);

      if ((servers[x]= memcached_create(NULL)) == NULL)
      {
        fprintf(stderr, "Failed to memcached_create()\n");
        if (x > 0)
          memcached_free(servers[0]);
        x--;

        for (; x > 0; x--)
          memcached_free(servers[x]);

        free(servers);
        return;
      }
      memcached_server_add(servers[x],
                           memcached_server_name(instance),
                           memcached_server_port(instance));
    }

    printf("Network Latency Test:\n\n");
    struct timeval start_time, end_time;
    uint32_t slowest_server= 0;
    long elapsed_time, slowest_time= 0;

    for (uint32_t x= 0; x < server_count; x++)
    {
      memcached_server_instance_st instance=
        memcached_server_instance_by_position(memc, x);
      gettimeofday(&start_time, NULL);

      for (uint32_t y= 0; y < num_of_tests; y++)
      {
        size_t vlen;
        char *val= memcached_get(servers[x], test_key, strlen(test_key),
                                 &vlen, &flags, &rc);
        if (rc != MEMCACHED_NOTFOUND && rc != MEMCACHED_SUCCESS)
          break;
        free(val);
      }
      gettimeofday(&end_time, NULL);

      elapsed_time= (long) timedif(end_time, start_time);
      elapsed_time /= (long) num_of_tests;

      if (elapsed_time > slowest_time)
      {
        slowest_server= x;
        slowest_time= elapsed_time;
      }

      if (rc != MEMCACHED_NOTFOUND && rc != MEMCACHED_SUCCESS)
      {
        printf("\t %s (%d)  =>  failed to reach the server\n",
               memcached_server_name(instance),
               memcached_server_port(instance));
      }
      else
      {
        printf("\t %s (%d)  =>  %ld.%ld seconds\n",
               memcached_server_name(instance),
               memcached_server_port(instance),
               elapsed_time / 1000, elapsed_time % 1000);
      }
    }

    if (server_count > 1 && slowest_time > 0)
    {
      memcached_server_instance_st slowest=
        memcached_server_instance_by_position(memc, slowest_server);

      printf("---\n");
      printf("Slowest Server: %s (%d) => %ld.%ld seconds\n",
             memcached_server_name(slowest),
             memcached_server_port(slowest),
             slowest_time / 1000, slowest_time % 1000);
    }
    printf("\n");

    for (uint32_t x= 0; x < server_count; x++)
      memcached_free(servers[x]);

    free(servers);
    free(analyze_mode);
  }
  else
  {
    fprintf(stderr, "Invalid Analyzer Option provided\n");
    free(analyze_mode);
  }
}

static void print_analysis_report(memcached_st *memc,
                                  memcached_analysis_st *report)
                                  
{
  uint32_t server_count= memcached_server_count(memc);
  memcached_server_instance_st most_consumed_server= memcached_server_instance_by_position(memc, report->most_consumed_server);
  memcached_server_instance_st least_free_server= memcached_server_instance_by_position(memc, report->least_free_server);
  memcached_server_instance_st oldest_server= memcached_server_instance_by_position(memc, report->oldest_server);

  printf("Memcached Cluster Analysis Report\n\n");

  printf("\tNumber of Servers Analyzed         : %u\n", server_count);
  printf("\tAverage Item Size (incl/overhead)  : %u bytes\n",
         report->average_item_size);

  if (server_count == 1)
  {
    printf("\nFor a detailed report, you must supply multiple servers.\n");
    return;
  }

  printf("\n");
  printf("\tNode with most memory consumption  : %s:%u (%llu bytes)\n",
         memcached_server_name(most_consumed_server),
         (uint32_t)memcached_server_port(most_consumed_server),
         (unsigned long long)report->most_used_bytes);
  printf("\tNode with least free space         : %s:%u (%llu bytes remaining)\n",
         memcached_server_name(least_free_server),
         (uint32_t)memcached_server_port(least_free_server),
         (unsigned long long)report->least_remaining_bytes);
  printf("\tNode with longest uptime           : %s:%u (%us)\n",
         memcached_server_name(oldest_server),
         (uint32_t)memcached_server_port(oldest_server),
         report->longest_uptime);
  printf("\tPool-wide Hit Ratio                : %1.f%%\n", report->pool_hit_ratio);
  printf("\n");
}

static void options_parse(int argc, char *argv[])
{
  memcached_programs_help_st help_options[]=
  {
    {0},
  };

  int option_index= 0;

  bool opt_version= false;
  bool opt_help= false;
  while (1) 
  {
    int option_rv= getopt_long(argc, argv, "Vhvds:a", long_options, &option_index);

    if (option_rv == -1)
      break;

    switch (option_rv)
    {
    case 0:
      break;

    case OPT_VERBOSE: /* --verbose or -v */
      opt_verbose= true;
      break;

    case OPT_DEBUG: /* --debug or -d */
      opt_verbose= true;
      break;

    case OPT_BINARY:
      opt_binary= true;
      break;

    case OPT_SERVER_VERSION:
      opt_server_version= true;
      break;

    case OPT_VERSION: /* --version or -V */
      opt_version= true;
      break;

    case OPT_HELP: /* --help or -h */
      opt_help= true;
      break;

    case OPT_SERVERS: /* --servers or -s */
      opt_servers= strdup(optarg);
      break;

    case OPT_STAT_ARGS:
      stat_args= strdup(optarg);
      break;

    case OPT_ANALYZE: /* --analyze or -a */
      opt_analyze= OPT_ANALYZE;
      analyze_mode= (optarg) ? strdup(optarg) : NULL;
      break;

    case OPT_QUIET:
      close_stdio();
      break;

    case '?':
      /* getopt_long already printed an error message. */
      exit(1);
    default:
      abort();
    }
  }

  if (opt_version)
  {
    version_command(PROGRAM_NAME);
    exit(EXIT_SUCCESS);
  }

  if (opt_help)
  {
    help_command(PROGRAM_NAME, PROGRAM_DESCRIPTION, long_options, help_options);
    exit(EXIT_SUCCESS);
  }
}
